File: Toggly.FeatureManagement.NET/Toggly.FeatureManagement/TogglySettings.cs
using System;
using System.Collections.Generic;

namespace Toggly.FeatureManagement
{
    public class TogglySettings
    {
        /// <summary>
        /// Toggly App Key. Get it from the App Settings page on toggly.io
        /// </summary>
        public string AppKey { get; set; } = string.Empty;

        /// <summary>
        /// Name of the environment. Case sensitive
        /// </summary>
        public string Environment { get; set; } = "Production";

        /// <summary>
        /// Use signed definitions to get feature updates.
        /// Strongly recommended in production when flags gate security-sensitive behavior.
        /// </summary>
        public bool UseSignedDefinitions { get; set; }

        /// <summary>
        /// Base URL for metrics/usage gRPC (trusted configuration). Defaults to https://app.toggly.io/.
        /// Only override for self-hosted or private deployments you control.
        /// </summary>
        public string? BaseUrl { get; set; }

        /// <summary>
        /// Base URL for definitions, WebSocket, and JWKS (trusted configuration).
        /// Defaults to https://definitions.toggly.io/. Only override for deployments you control;
        /// a malicious value can redirect fetch/JWKS traffic (SSRF-style misconfiguration risk).
        /// </summary>
        public string? DefinitionsBaseUrl { get; set; }

        /// <summary>
        /// The current version of the application. Used to track deployments
        /// Assembly version is used if not specified
        /// </summary>
        public string? AppVersion { get; set; }

        /// <summary>
        /// Hostname or instance name of the application. Useful in load-balanced and multi-server setups
        /// Hostname is used if not specified
        /// </summary>
        public string? InstanceName { get; set; }

        /// <summary>
        /// Undefined features should be treated as AlwaysOn on development
        /// (when app.Environment.IsDevelopment() is true)
        /// </summary>
        public bool UndefinedEnabledOnDevelopment { get; set; }

        /// <summary>
        /// Optional whitelist of JWKS key IDs trusted for signed definitions.
        /// </summary>
        public HashSet<string>? AllowedKeyIds { get; set; }

        /// <summary>
        /// How long verified JWKS keys are cached in memory and in the snapshot store.
        /// Default is 30 days. Minimum effective value is 1 minute.
        /// </summary>
        public TimeSpan JwksCacheDuration { get; set; } = TimeSpan.FromDays(30);

        /// <summary>
        /// Optional callback invoked when fetch, cache, signature, JWKS, or storage
        /// operations fail. Use for Sentry/AppInsights or other telemetry.
        /// The last error is also available via <see cref="IFeatureProviderDebug.GetDebugInfo"/>.
        /// </summary>
        public Action<string, Exception?>? OnError { get; set; }

        /// <summary>
        /// When true (default), discovered entity context schemas are registered with Toggly at startup.
        /// Failures are logged and do not block application startup.
        /// </summary>
        public bool RegisterContextsOnStartup { get; set; } = true;
    }
}

File: Toggly.FeatureManagement.PHP/src/Toggly/FeatureManagement/Config/TogglySettings.php
<?php

namespace Toggly\FeatureManagement\Config;

/**
 * Configuration settings for Toggly Feature Management
 */
class TogglySettings
{
    /**
     * Toggly App Key. Get it from the App Settings page on toggly.io
     */
    public string $appKey = '';

    /**
     * Name of the environment. Case sensitive
     */
    public string $environment = 'Production';

    /**
     * Use signed definitions to get feature updates
     */
    public bool $useSignedDefinitions = false;

    /**
     * Base URL of the toggly instance. Leave blank unless you have a reason to change
     */
    public ?string $baseUrl = null;

    /**
     * The current version of the application. Used to track deployments
     */
    public ?string $appVersion = null;

    /**
     * Hostname or instance name of the application. Useful in load-balanced and multi-server setups
     */
    public ?string $instanceName = null;

    /**
     * Undefined features should be treated as AlwaysOn on development
     */
    public bool $undefinedEnabledOnDevelopment = false;

    /**
     * Whitelist of allowed key IDs for signed definitions
     * @var string[]|null
     */
    public ?array $allowedKeyIds = null;

    /**
     * Refresh interval in seconds (default: 300 = 5 minutes)
     */
    public int $refreshInterval = 300;

    /**
     * Enable live updates via WebSocket (default: true)
     * Only effective in long-running processes (CLI, Laravel Octane, ReactPHP, etc.)
     * Automatically disabled in PHP-FPM / traditional request-per-process environments
     */
    public bool $enableLiveUpdates = true;

    public function __construct(array $config = [])
    {
        if (isset($config['app_key'])) {
            $this->appKey = $config['app_key'];
        }
        if (isset($config['environment'])) {
            $this->environment = $config['environment'];
        }
        if (isset($config['use_signed_definitions'])) {
            $this->useSignedDefinitions = (bool)$config['use_signed_definitions'];
        }
        if (isset($config['base_url'])) {
            $this->baseUrl = $config['base_url'];
        }
        if (isset($config['app_version'])) {
            $this->appVersion = $config['app_version'];
        }
        if (isset($config['instance_name'])) {
            $this->instanceName = $config['instance_name'];
        }
        if (isset($config['undefined_enabled_on_development'])) {
            $this->undefinedEnabledOnDevelopment = (bool)$config['undefined_enabled_on_development'];
        }
        if (isset($config['allowed_key_ids'])) {
            $this->allowedKeyIds = is_array($config['allowed_key_ids']) 
                ? $config['allowed_key_ids'] 
                : explode(',', $config['allowed_key_ids']);
        }
        if (isset($config['refresh_interval'])) {
            $this->refreshInterval = (int)$config['refresh_interval'];
        }
        if (isset($config['enable_live_updates'])) {
            $this->enableLiveUpdates = (bool)$config['enable_live_updates'];
        }
    }

    /**
     * Get the base URL, defaulting to https://definitions.toggly.io/
     */
    public function getBaseUrl(): string
    {
        return $this->baseUrl ?? 'https://definitions.toggly.io/';
    }
}
